util/fork_func.c

```diff
@@ -0,0 +1,77 @@
+#include <sys/wait.h>
+
+#include <stdint.h>
+#include <unistd.h>
+
+#include "warnp.h"
+
+#include "fork_func.h"
+
+/**
+ * fork_func(func, cookie):
+ * Fork and run ${func} in a new process, with ${cookie} as the sole argument.
+ */
+pid_t
+fork_func(int (* func)(void *), void * cookie)
+{
+	pid_t pid;
+
+	/* Fork */
+	if ((pid = fork()) == -1) {
+		warnp("fork");
+		goto err0;
+	}
+	if (pid == 0) {
+		/* In child process: Run the provided function, then exit. */
+		_exit((*func)(cookie));
+	}
+
+	/* In parent process: do nothing else. */
+
+	/* Success! */
+	return (pid);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * fork_func_wait(pid):
+ * Wait for the process ${pid} to finish.  Print any error arising from ${pid}.
+ * If ${pid} exited cleanly, return its exit code; otherwise, return -1.
+ */
+int
+fork_func_wait(pid_t pid)
+{
+	int status;
+	int rc = -1;
+
+	/* Wait for the process to finish. */
+	if (waitpid(pid, &status, 0) == -1) {
+		warnp("waitpid");
+		goto err0;
+	}
+
+	/* Print the error status, if applicable. */
+	if (WIFEXITED(status)) {
+		/* Child ${pid} exited cleanly. */
+		rc = WEXITSTATUS(status);
+	} else {
+		/* Child ${pid} did not exit cleanly. */
+
+		/* Warn about the reason for the unclean exit. */
+		if (WIFSIGNALED(status))
+			warn0("pid %jd: signalled with %i",
+			    (intmax_t)pid, WTERMSIG(status));
+		else if (WIFSTOPPED(status))
+			warn0("pid %jd: stopped with %i",
+			    (intmax_t)pid, WSTOPSIG(status));
+		else
+			warn0("pid %jd: stopped for unknown reason.");
```

Comment:
Is "stopped" what you want here?  I mean, stopped != exited.

Reply:
Good catch, changed.